core/primitives/src/epoch_manager.rs

```diff
@@ -46,6 +49,55 @@ pub struct EpochConfig {
     pub protocol_upgrade_stake_threshold: Rational,
     /// Number of epochs after stake threshold was achieved to start next prtocol version.
     pub protocol_upgrade_num_epochs: EpochHeight,
+    /// Shard layout of this epoch, may change from epoch to epoch
+    pub shard_layout: ShardLayout,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ShardConfig {
+    pub num_block_producer_seats_per_shard: Vec<NumSeats>,
+    pub avg_hidden_validator_seats_per_shard: Vec<NumSeats>,
+    pub shard_layout: ShardLayout,
+}
+
+#[derive(Clone)]
+pub struct AllEpochConfig {
+    genesis_epoch_config: Arc<EpochConfig>,
+    simple_nightshade_epoch_config: Arc<EpochConfig>,
+}
+
+impl AllEpochConfig {
+    pub fn new(
+        genesis_epoch_config: EpochConfig,
+        simple_nightshade_shard_config: Option<ShardConfig>,
+    ) -> Self {
+        let mut config = genesis_epoch_config.clone();
+        if let Some(ShardConfig {
+            num_block_producer_seats_per_shard,
+            avg_hidden_validator_seats_per_shard,
+            shard_layout,
+        }) = simple_nightshade_shard_config
+        {
+            config.num_block_producer_seats_per_shard = num_block_producer_seats_per_shard;
+            config.avg_hidden_validator_seats_per_shard = avg_hidden_validator_seats_per_shard;
+            config.shard_layout = shard_layout;
+        }
```

Comment:
```suggestion
        let config = match simple_nightshade_shard_config {
            Some(config) => config.clone(),
            None => genesis_epoch_config.clone()
        };
```

Reply:
ShardConfig is different from EpochConfig here. simple_nightshade_shard_config is a 'ShardConfig` and `config` is a `EpochConfig`